First stab at figuring out the number of very active mentors

MaineC opened the pull request:
Relates to #114

This is a work in progress pull request. The goal is to extract the number of very active mentors in a project based on number of comments they left in issues and pull requests.

In order to avoid including endless discussion threads that often go in circles the idea is to limit the number of issue comments and pr comments taken into consideration.

This is also where I need input:

As in many other issue-metrics stats I used https://github3.readthedocs.io/en/1.0.1/issues.html#github3.issues.issue.Issue.comments in order to iterate over comments. In the documentation it says the iterator can limit the number of comments considered. In my test case though it seems to ignore this parameter. Likely I mis-understood the use of the iterator?

## Readiness Checklist

### Author/Contributor
- [ ] If documentation is needed for this change, has that been included in this pull request
- [x] run `make lint` and fix any issues that you have introduced
- [x] run `make test` and ensure you have test coverage for the lines you are introducing

### Reviewer
- [x] Label as either `bug`, `documentation`, `enhancement`, `infrastructure`, or `breaking`

zkoppert commented on most_active_mentors.py:
```diff
@@ -0,0 +1,149 @@
+
+Open questions:
+   - should there be an option to limit this to certain users, e.g. core maintainers?
+   - should there be a limit to how many comments per PR we consider to avoid having
```
Yes, a limit makes sense to me. It may make sense to say that after 1 comment someone was involved with a PR, and after 3 comments they were heavily involved, and to stop counting after that. Those numbers are my personal thresholds in my mind but could be set to configurable numbers with reasonable defaults.

zkoppert commented on most_active_mentors.py:
```diff
@@ -0,0 +1,149 @@
+   - should there be an option to limit this to certain users, e.g. core maintainers?
+   - should there be a limit to how many comments per PR we consider to avoid having
+     the statistic dominated by contested PRs?
+   - should this metric count consecutive comments coming from the same user as only
```
Great question. In this case I would also consider the length of the comment. If somebody has a lot to say and they break that up into several meaningful comments, I would see that as several contributions worth counting.

zkoppert reviewed (commented):
I see this as really valuable work! Thanks for working on it! ✨ 

For larger projects or organizations, it would be especially great to be able to see this data to be able to quickly find active projects that lack mentorship.

zkoppert commented:
next step for this one looks like working on fixing the tests.

MaineC commented:
I fixed the linter errors, ran the tests.

Still on my TODO list: 
* Option to enable/disable this check.
* Output of results.

MaineC commented:
Output added to markdown and call added - and fixed the issues with the original design.

Still missing:
* Output in json.
* Additional configuration options for disabling this feature and configuring e.g. cutoff for who is considered an active mentor.

MaineC commented:
In the last commit I added mentor counting results to json output. In addition max number of comments per issue as well as the cutoff number for heavily involved mentors are now configurable.

Tests seem to run through locally, including the ones I added.

`make lint` left two warnings.

With that I think this is ready for getting a closer look and feedback.

Wish you Happy Easter (no, no Easter eggs hidden here ;) )

MaineC commented:
Fixed merging errors.

There are two lint warnings still open that I can't figure out.

zkoppert commented:
I can dig into some of these linting errors. I'll set aside some time this coming week.

zkoppert commented:
@MaineC I was able to correct the linting errors and override some unreasonable linting rules. Ready to click the ready for review button?

MaineC commented:
Clicked the button.

jmeridth commented on config.py:
```diff
@@ -51,11 +55,16 @@ def __init__(
         self.gh_app_id = gh_app_id
         self.gh_app_installation_id = gh_app_installation_id
         self.gh_app_private_key_bytes = gh_app_private_key_bytes
+        self.search_query = search_query
```
already on line 80 below

MaineC replied:
Likely a residual of me fixing mering errors. Removed.

jmeridth commented on issue_metrics.py:
```diff
@@ -307,6 +333,10 @@ def main():
     stats_time_to_answer = get_stats_time_to_answer(issues_with_metrics)
 
+    num_mentor_count = 0
+    if enable_mentor_count == "TRUE":
```
```suggestion
    if enable_mentor_count:
```

my suggestion in config.py L186 will allow this to be a boolean and this can now be above code

jmeridth commented on requirements.txt:
```diff
@@ -1,3 +1,4 @@
+flake8==7.0.0
```
flake8 requirement is in `requirements-test.txt`

MaineC replied:
Ah - I see. I hadn't noticed that one.

jmeridth commented on test_issue_metrics.py:
```diff
@@ -123,7 +123,7 @@ def test_get_env_vars_missing_query(self):
         # Call the function and check that it raises a ValueError
         with self.assertRaises(ValueError):
-            get_env_vars()
+            get_env_vars(test=True)
```
nice catch, thank you

jmeridth commented on test_most_active_mentors.py:
```diff
@@ -0,0 +1,82 @@
+        """Test that get_mentor_count correctly counts comments per user."""
+        mentor_activity = {"sue": 15, "bob": 10}
+
+        # Create moc data
```
```suggestion
        # Create mock data
```

jmeridth reviewed (changes requested):
Looks good.  Just a couple asks.  Thank you for your contribution.

zkoppert commented:
@MaineC Feel free to review and integrate the change requests from @jmeridth.

jmeridth commented on config.py:
```diff
@@ -34,6 +34,10 @@ class EnvVars:
         hide_time_to_first_response (bool): If true, the time to first response metric is hidden in the output
         ignore_users (List[str]): List of usernames to ignore when calculating metrics
         labels_to_measure (List[str]): List of labels to measure how much time the lable is applied
+        enable_mentor_count (str): If set to TRUE, compute number of mentors
```
Needs to be `bool` now with our other changes.

MaineC replied:
True. That's what I get for only integrating your suggestions through the web UI.

jmeridth replied:
We've all done it. 😄

jmeridth commented on config.py:
```diff
@@ -51,6 +55,10 @@ def __init__(
         hide_time_to_first_response: bool,
         ignore_user: List[str],
         labels_to_measure: List[str],
+        enable_mentor_count: str,
```
```suggestion
        enable_mentor_count: bool,
```

gotta change this too. 😄
